src/main/groovy/com/vanniktech/android/junit/jacoco/GenerationPlugin.groovy

```diff
@@ -139,85 +144,124 @@ class GenerationPlugin implements Plugin<Project> {
-            final def taskName = "jacocoTestReport${sourceName.capitalize()}"
+            final def jvmTaskName = "jacocoTestReport${sourceName.capitalize()}"
+            final def combinedTaskName = "combinedTestReport${sourceName.capitalize()}"
 
-            subProject.task(taskName, type: JacocoReport, dependsOn: testTaskName) {
-                group = 'Reporting'
-                description = "Generate Jacoco coverage reports after running ${sourceName} tests."
+            final def jvmTestTaskName = "test${sourceName.capitalize()}UnitTest"
+            final def instrumentationTestTaskName = "create${sourceName.capitalize()}CoverageReport"
 
-                reports {
-                    xml {
-                        enabled = true
-                        destination subProject.file("${subProject.buildDir}/reports/jacoco/${sourceName}/jacoco.xml")
-                    }
-                    csv {
-                        enabled = true
-                        destination subProject.file("${subProject.buildDir}/reports/jacoco/${sourceName}/jacoco.csv")
-                    }
-                    html {
-                        enabled = true
-                        destination subProject.file("${subProject.buildDir}/reports/jacoco/${sourceName}")
-                    }
-                }
+            addJacocoTask(false, subProject, extension, mergeTask, mergedReportTask, jvmTaskName,
+                jvmTestTaskName, instrumentationTestTaskName, sourceName, sourcePath, productFlavorName, buildTypeName)
 
-                def classPaths = [
-                  "**/intermediates/classes/${sourcePath}/**",
-                  "**/intermediates/javac/${sourceName}/*/classes/**" // Android Gradle Plugin 3.2.x support.
-                ]
+            if (extension.configureInstrumentationCoverage && buildType.testCoverageEnabled) {
+                addJacocoTask(true, subProject, extension, mergeTask, mergedReportTask, combinedTaskName,
+                    jvmTestTaskName, instrumentationTestTaskName, sourceName, sourcePath, productFlavorName, buildTypeName)
+            }
+        }
 
-                if (isKotlinAndroid(subProject)) {
-                    classPaths << "**/tmp/kotlin-classes/${sourcePath}/**"
-                    if (productFlavorName) {
-                        classPaths << "**/tmp/kotlin-classes/${productFlavorName}${buildTypeName.capitalize()}/**"
-                    }
+        return true
+    }
+
+    private static void addJacocoTask(final boolean combined, final Project subProject, final JunitJacocoExtension extension,
+                                      JacocoMerge mergeTask, JacocoReport mergedReportTask, final String taskName,
+                                      final String jvmTestTaskName, final String instrumentationTestTaskName, final String sourceName,
+                                      final String sourcePath, final String productFlavorName, final String buildTypeName) {
```

Comment:
It's ugly to pass all these arguments, but I wanted to make the task creation reusable.